### ISSUE TITLE
make the pattern the primary object for args

### ISSUE THREAD

**Issue by davidmarkclements** [apparatus/mu#43]
since accessing the pattern is definitely the main case, what if we make it the 
args object, and make protocol accessible either through an additional parameter or on the proto of the args object. From the transport, given we have an object (`o`) that is composed {pattern, protocol} we can do

```js
var args = o.pattern
pattern.__proto__ = {protocol: protocol}
```

Then we would use like so

```js
murm.define({role: 'error', cmd: 'error'}, function (args, cb) {
   console.log(args.role === 'error') // true
   console.log(args.protocol.ttl) // 10
})
```

since protocol is on the proto, it won't show up during serialization (or logging, but we can make a pino serializer)

The other approach would be 

```js
murm.define({role: 'error', cmd: 'error'}, function (args, cb, protocol) {
   console.log(args.role === 'error') // true
   console.log(protocol.ttl) // 10
})
```

Or we could go polymorphic with `(args, cb)` and `(args, protocol, cb)` being appropriately handled

cc @pelger @mcollina

**Comment by davidmarkclements** [comment]
In either case we can also add pattern to the proto to ensure backwards compatible

**Comment by mcollina** [comment]
I am missing some parts here: how `protocol`  ends up within the pattern?
From this example, this seems not the case: https://github.com/apparatus/mu/blob/master/test/core/basic.test.js#L31-L34.

I do not think that touching the prototype makes sense. The best way to handle this is with the use of `Symbol`.

The big question is, do we need access this metadata in the browser?

**Comment by davidmarkclements** [comment]
yah so what I'm saying, is that we have an args object with a pattern property

so you have to access args.pattern

say pattern is {a; 1, b: 2}

instead of args.pattern.a

I want args.a

and then we include the protocol metadata as a non-enumerable `args.protocol` - the other option here is Object.defineProperty as non-enumerable, doesn't have to be proto at all now I think about it

for browser,  once we want to do network transports in the browser (http, websocket, webrtc), we're going to need protocol meta data. there may even be a reason for local transports too

**Comment by mcollina** [comment]
I'm missing something. Given the above test, we have:

`mu.dispatch(pattern)`
`mu.define(pattern, function (pattern, done) {})`

Maybe I am missing something

**Comment by davidmarkclements** [comment]
@mcollina 

no, the pattern goes into an `args` object alongside `protocol`

```js
mu.dispatch(pattern)
mu.define(pattern, function ({protocol, pattern}, done) {})
```

doing the boilerplate stuff now has me convinced this needs to change, 
it's counterintuitive, too easy to leak protocol info to a user consider this:

```js
mu.define({role: 'my-service', cmd: 'do-thing'}, (args, cb) => {
  mu.dispatch({role: 'another-service', cmd: 'action'}, (err, result) => {
    if (err) return cb(err)
    args.moreStuff = result
    cb(null, args)
  })
})
```

calling `{role: 'my-service', cmd: 'do-thing'}` elsewhere will result in:

```
{"pattern":{role: "my-service", cmd: "do-thing"},"protocol":{"path":["a15d645b-a269-40a7-826f-e0b8628026c9","56b46755-86ea-4836-845c-26b33a0770f0"],"trace":["a15d645b-a269-40a7-826f-e0b8628026c9","56b46755-86ea-4836-845c-26b33a0770f0"],"ttl":10,"dst":"target","src":"56b46755-86ea-4836-845c-26b33a0770f0","inboundIfc":"ddd71529-3a0b-45af-b8b2-2970ed07c271"}, "moreStuff": {"wtf": "am i doing here?"}}
```

what you have to do instead is 

```js
    args.pattern.moreStuff = result
    cb(null, args.pattern)
```

this is pretty big gotcha, it we currently allow the user to throw data outside of both pattern and protocol

cc @pelger

**Comment by davidmarkclements** [comment]
I've also realised that (of course!) we don't have to do the whole proto thing, we can just set protocol as non-enumerable on pattern, and pattern becomes args

**Comment by mcollina** [comment]
We can also use symbols, but it will make things harder in browsers.
